gather.go

```diff
@@ -90,7 +90,38 @@ func listenUDP(portMax, portMin int, network string, laddr *net.UDPAddr) (*net.U
 	return nil, ErrPort
 }
 
-func gatherCandidatesLocal(a *Agent, networkTypes []NetworkType) {
+// GatherCandidates initiates the trickle based gathering process.
+func (a *Agent) GatherCandidates(urls []*URL, networkTypes []NetworkType) error {
+	return a.run(func(agent *Agent) {
+		if a.gatheringState == GatheringStateGathering {
+			a.log.Warnf("Attempting to gather candidates during gathering state\n")
+			return
+		}
+
+		go a.gatherCandidates(&AgentConfig{
+			Urls:         urls,
+			NetworkTypes: networkTypes,
+		})
+	})
+}
+
+func (a *Agent) gatherCandidates(config *AgentConfig) {
+	a.gatheringState = GatheringStateGathering
```

Comment:
I think we need to lock while setting this.

IMO I think it would be nice to have this as an atomic. Just add a getter/setter and then we don't need to worry :)

Reply:
I can close this stuff behind `a.run`. You good with that?

Reply:
Yea that is good with me!

Reply:
Fixed!